pkg/controller/notification/mailgun_notification_delivery_service.go

```diff
@@ -2,17 +2,96 @@ package notification
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"github.com/mailgun/mailgun-go/v4"
+	"time"
 )
 
+type MailgunDeliveryError struct {
+	id           string
+	response     string
+	errorMessage string
+}
+
+func (e MailgunDeliveryError) Error() string {
+	return fmt.Sprintf("error while delivering notification (ID: %s, Response: %s) - %s", e.id, e.response, e.errorMessage)
+}
+
+func NewMailgunDeliveryError(id, response, errorMessage string) error {
+	return MailgunDeliveryError{
+		id:           id,
+		response:     response,
+		errorMessage: errorMessage,
+	}
+}
+
 type MailgunNotificationDeliveryService struct {
+	base        BaseNotificationDeliveryService
+	Mailgun     mailgun.Mailgun
+	SenderEmail string
 }
 
-// NewMailgunNotificationDeliveryService
-func NewMailgunNotificationDeliveryService() NotificationDeliveryService {
-	return &MailgunNotificationDeliveryService{}
+type MailgunConfiguration interface {
+	GetMailgunDomain() string
+	GetMailgunAPIKey() string
+	GetMailgunSenderEmail() string
 }
 
-func (s *MailgunNotificationDeliveryService) Send(ctx context.Context, notificationCtx NotificationContext, templateName string) error {
-	// TODO implement
+type MailgunOption interface {
+	// ApplyToMailgun applies this configuration to the given mailgun instance.
+	ApplyToMailgun(mailgun.Mailgun)
+}
+
+// NewMailgunNotificationDeliveryService creates a delivery service that uses the Mailgun API to deliver email notifications
+func NewMailgunNotificationDeliveryService(config MailgunConfiguration, templateLoader TemplateLoader,
+	opts ...MailgunOption) NotificationDeliveryService {
+
+	mg := mailgun.NewMailgun(config.GetMailgunDomain(), config.GetMailgunAPIKey())
```

Comment:
not sure if we want to do a check here but if the config or templateLoader ends up being nil this will panic.

Reply:
Not sure if it's worth doing a check here, we can guarantee these parameters are not nil elsewhere.